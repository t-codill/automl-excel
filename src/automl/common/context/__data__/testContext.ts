import { PageNames } from "../../PageNames";
import { Logger } from "../../utils/logger";
import { IWorkspaceContextProps, WorkspaceError } from "../IWorkspaceProps";
import { WorkspaceFlight } from "../WorkspaceFlight";

export function getLogCustomEventSpy(): jest.SpyInstance<ReturnType<Logger["logCustomEvent"]>> {
    const spy = jest.spyOn(Logger.prototype, "logCustomEvent");
    spy.mockImplementation(jest.fn());
    return spy;
}
let pageName: PageNames = PageNames.Unknown;
export const testContext: IWorkspaceContextProps = {
    get pageName(): PageNames { return pageName; },
    get theme(): string { return "light"; },
    discoverUrls: {},
    flight: new WorkspaceFlight(""),
    location: "eastus",
    resourceGroupName: "testResource",
    subscriptionId: "00000000-0000-0000-0000-000000000000",
    logger: new Logger("test"),
    getToken: () => "testToken",
    setPageName: (value: PageNames) => { pageName = value; },
    workspaceName: "testWorkSpace",
    setNavigationBarButtons(): void { return; },
    onError: (err: WorkspaceError) => { throw err; },
    clearErrors: () => { return; },
    setHeader: () => { return; },
    setLoading: () => { return; }
};
