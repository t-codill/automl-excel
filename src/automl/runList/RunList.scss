.run-filter-container {
    .run-filter-action-container {
        width: 300px;
        display: inline-block;
        margin-left: 16px;
    }
}

.run-status-container {
    display: flex;
    width: 100%;
    padding-top: 1em;

    .run-status-card-container {
        width: 400px;
        flex: 0 0 auto;
    }

    .run-status-chart-container {
        flex: 1 0 auto;
    }
}