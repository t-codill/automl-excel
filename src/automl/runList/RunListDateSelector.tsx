import { values } from "lodash";
import { Dropdown, IDropdownOption } from "office-ui-fabric-react";
import * as React from "react";

interface IRunListDateSelectorProps {
    onDateChange(
        startDate: Date | undefined
    ): void;
}

interface IRunListDateSelectorState {
    dateSelector: string | number | undefined;
    dropdownOptions: Array<{ key: string; text: string }>;
}

export const dateFilterDropdownOptions = {
    ALL: {
        key: "allDate",
        text: "All dates"
    },
    THIRTYDAYS: {
        key: "30days",
        text: "Last 30 days"
    },
    SIXTYDAYS: {
        key: "60days",
        text: "Last 60 days"
    },
    NINTYDAYS: {
        key: "90days",
        text: "Last 90 days"
    }
};

export class RunListDateSelector extends React.Component<IRunListDateSelectorProps, IRunListDateSelectorState
    > {
    constructor(props: IRunListDateSelectorProps) {
        super(props);
        const dropdownOptions = values(dateFilterDropdownOptions);
        this.state = {
            dateSelector: dateFilterDropdownOptions.ALL.key,
            dropdownOptions,
            // showDatePicker: false
        };
    }
    public readonly render = (): React.ReactNode => {
        return (
            <>
                <Dropdown
                    placeholder="All Dates"
                    onChange={this.changeState}
                    selectedKey={this.state.dateSelector}
                    options={this.state.dropdownOptions}
                />
            </>
        );
    }
    public changeState = (_event: React.FormEvent<HTMLDivElement>, option?: IDropdownOption): void => {
        this.setState((prevState) => {
            if (option) {
                if (option.key !== this.state.dateSelector) {
                    let diff = 0;
                    switch (option.key) {
                        case dateFilterDropdownOptions.THIRTYDAYS.key:
                            diff = -30;
                            break;
                        case dateFilterDropdownOptions.SIXTYDAYS.key:
                            diff = -60;
                            break;
                        case dateFilterDropdownOptions.NINTYDAYS.key:
                            diff = -90;
                            break;
                        default:
                            diff = 0;
                    }
                    const startDate = new Date();
                    startDate.setDate(startDate.getDate() + diff);
                    this.props.onDateChange(diff === 0 ? undefined : startDate);
                    return { dateSelector: option.key };
                }
            }
            return { dateSelector: prevState.dateSelector };
        });
    }
}
