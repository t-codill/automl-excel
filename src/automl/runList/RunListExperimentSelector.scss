.experiment-selector-callout {
    .ms-Dropdown-header {
        height: 40px;
    }
}