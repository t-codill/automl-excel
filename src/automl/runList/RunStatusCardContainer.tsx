import { Dictionary } from "lodash";
import { FocusZone } from "office-ui-fabric-react";
import * as React from "react";
import { RunStatusCard } from "./RunStatusCard";

const mergedStatus: string[] = ["Running", "Completed", "Failed", "Others"];

interface IRunStatusCardContainerProps {
    runStatus: Dictionary<number>;
    onStatusFilterChange(
        statusFilter: string | undefined
    ): void;
}
interface IRunStatusCardContainerState {
    selectedCard: string | undefined;
}
export class RunStatusCardContainer extends React.Component<IRunStatusCardContainerProps, IRunStatusCardContainerState>{
    constructor(props: IRunStatusCardContainerProps) {
        super(props);
        this.state = {
            selectedCard: undefined
        };
    }
    public readonly render = (): React.ReactNode => {
        return (
            <FocusZone className="ms-Grid">
                {mergedStatus.map((s) =>
                    <RunStatusCard key={s}
                        className="ms-Grid-col ms-sm6 ms-md6 ms-lg6 ms-xl6 ms-xxl6 ms-xxxl6"
                        isActive={this.state.selectedCard === s}
                        handleStatusFilterChange={this.handleStatusCardClick}
                        runStatusLabel={s}
                        runStatusCount={this.props.runStatus[s] || 0} />
                )}
            </FocusZone>
        );
    }
    private readonly handleStatusCardClick = (selectedCard: string | undefined) => {
        this.setState({ selectedCard });
        this.props.onStatusFilterChange(selectedCard);
    }
}
