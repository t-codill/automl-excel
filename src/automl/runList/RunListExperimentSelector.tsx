
import { map } from "lodash";
import { Dropdown, DropdownMenuItemType, IDropdownOption, Toggle } from "office-ui-fabric-react";
import * as React from "react";

import "./RunListExperimentSelector.scss";

interface IExperimentSelectorOption {
    key: string;
    text: string;
}

interface IRunListExperimentSelectorProps {
    experimentNames: string[] | undefined;
    isAllSelect: boolean;
    onExperimentChange(
        experiments: string[] | undefined
    ): void;
}

interface IRunListExperimentSelectorState {
    availableOptions: IExperimentSelectorOption[];
    experimentSelectedItems?: string[];
    isAllSelect: boolean;
    experimentNames: string[];
}

export class RunListExperimentSelector extends React.Component<IRunListExperimentSelectorProps, IRunListExperimentSelectorState
    > {
    constructor(props: IRunListExperimentSelectorProps) {
        super(props);
        const availableOptions = this.getOptions(this.props.experimentNames);
        const experimentNames = map(availableOptions.slice(2), (option) => option.key);
        this.state = {
            availableOptions,
            experimentSelectedItems: experimentNames,
            isAllSelect: this.props.isAllSelect,
            experimentNames
        };
    }
    public readonly render = (): React.ReactNode => {
        return (
            <>
                <Dropdown
                    calloutProps={{
                        className: "experiment-selector-callout"
                    }}
                    selectedKeys={this.state.experimentSelectedItems}
                    multiSelect={true}
                    onChange={this.changeState}
                    placeholder="Select Experiments"
                    options={this.state.availableOptions}
                    onRenderOption={this.onRenderOption}
                    onRenderTitle={this.onRenderTitle}
                />
            </>
        );
    }
    private readonly onRenderTitle = () => {
        if (this.state.experimentSelectedItems && this.state.experimentSelectedItems.length === this.state.experimentNames.length) {
            return <span>All Experiments</span>;
        } else if (this.state.experimentSelectedItems && this.state.experimentSelectedItems.length === 1) {
            return <span>{this.state.experimentSelectedItems[0]}</span>;
        } else {
            return <span>{this.state.experimentSelectedItems && this.state.experimentSelectedItems.length} Experiments</span>;
        }
    }

    private readonly changeState = (_event: React.FormEvent<HTMLDivElement>, option?: IDropdownOption): void => {
        if (option) {
            this.setState((preState) => {
                const updatedSelectedItem = preState && preState.experimentSelectedItems ? [...preState.experimentSelectedItems] : [];
                if (option.selected) {
                    // add the option if it's checked
                    updatedSelectedItem.push(option.key.toString());
                } else {
                    // remove the option if it's unchecked
                    const currIndex = updatedSelectedItem.indexOf(option.key.toString());
                    updatedSelectedItem.splice(currIndex, 1);
                }
                this.props.onExperimentChange(updatedSelectedItem);
                return {
                    experimentSelectedItems: updatedSelectedItem,
                    isAllSelect: updatedSelectedItem.length === preState.experimentNames.length
                };
            });
        }
    }

    private readonly onToggleChanged = (_event: React.MouseEvent<HTMLElement, MouseEvent>, checked?: boolean | undefined) => {
        this.setState((preState) => {
            const updatedSelectedItem = checked ? preState.experimentNames : [];
            this.props.onExperimentChange(updatedSelectedItem);
            return {
                isAllSelect: !preState.isAllSelect,
                experimentSelectedItems: updatedSelectedItem
            };
        });
    }

    private readonly getOptions = (experimentNames: string[] | undefined) => {
        if (!experimentNames) {
            return [];
        }
        return [
            {
                key: "allOption",
                text: "Select All",
                itemType: DropdownMenuItemType.Header
            },
            {
                key: "divider",
                text: "-",
                itemType: DropdownMenuItemType.Divider
            },
            ...experimentNames.map((experimentName) => {
                return {
                    key: experimentName,
                    text: experimentName
                };
            })];
    }

    private readonly onRenderOption = (option?: IDropdownOption, defaultRender?: (props?: IDropdownOption) => JSX.Element | null) => {
        if (!option) {
            return <></>;
        }
        if (option.index === 0) {
            return (
                <Toggle
                    checked={this.state.isAllSelect}
                    inlineLabel={true}
                    label="Select All Experiment"
                    onChange={this.onToggleChanged} />
            );
        }
        if (defaultRender) {
            return defaultRender(option);
        }
        return <></>;
    }
}
