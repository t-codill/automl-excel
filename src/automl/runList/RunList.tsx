//import { RunHistoryChart } from "@mlworkspace/charts";
import { countBy, Dictionary, includes, map, uniq } from "lodash";
import moment from "moment";
import { ICommandBarItemProps } from "office-ui-fabric-react";
import * as React from "react";
import { ChildRun } from "../childRun/ChildRun";
import { PageNames } from "../common/PageNames";
import { processRunHistoryChart } from "../common/utils/processRunHistoryChart";
import { mergeRunStatus, statusMapping } from "../common/utils/run";
import { BasePage } from "../components/Base/BasePage";
import { DataTable } from "../components/DataTable/DataTable";
import { PageLoadingSpinner } from "../components/Progress/PageLoadingSpinner";
import { PageRedirect, PageRedirectLink } from "../components/Redirect/PageRedirect";
import { ParentRun } from "../parentRun/ParentRun";
import { AccessService } from "../services/AccessService";
import { IRunDtoWithExperimentName, RunHistoryService } from "../services/RunHistoryService";
import { StartRun } from "../startRun/StartRun";
import { Welcome } from "../welcome/Welcome";
import { RunListDateSelector } from "./RunListDateSelector";
import { RunListExperimentSelector } from "./RunListExperimentSelector";
import { RunStatusCardContainer } from "./RunStatusCardContainer";

import "./RunList.scss";

export interface IRunListState {
    runs: IRunDtoWithExperimentName[] | undefined;
    filteredRuns: IRunDtoWithExperimentName[] | undefined;
    uniqExperimentNames: string[] | undefined;
    dateFilter?: Date | undefined;
    experimentFilter?: string[] | undefined;
    statusFilter?: string | undefined;
    redirectToStartRun: boolean;
    runStatistics: Dictionary<number>;
    runHistoryChartData: Dictionary<Dictionary<number>>;
}
export class RunList extends BasePage<{}, IRunListState,
    { runHistoryService: RunHistoryService; accessService: AccessService }
    > {
    public static routePath = "";
    protected header: string | undefined = "Automated Machine Learning Dashboard";
    protected pageName = PageNames.RunList;
    protected readonly noBackButton = true;
    protected readonly serviceConstructors = { runHistoryService: RunHistoryService, accessService: AccessService };
    protected readonly navigationBarButtons: ICommandBarItemProps[] = [
        {
            key: "startRun",
            text: "Create experiment",
            iconProps: {
                iconName: "Add"
            },
            disabled: true,
            onClick: () => {
                this.startNewRun();
            }
        }
    ];

    public render(): React.ReactNode {
        if (!this.state || !this.state.runs) {
            return <PageLoadingSpinner />;
        }
        if (this.state.redirectToStartRun) {
            return PageRedirect(StartRun, {});
        }
        if (this.state.runs.length === 0) {
            return PageRedirect(Welcome, {});
        }
        return (<>
            <div className="run-filter-container">
                <div className="run-filter-action-container">
                    <RunListDateSelector onDateChange={this.handleDateFilterChange} />
                </div>
                <div className="run-filter-action-container">
                    <RunListExperimentSelector
                        isAllSelect={true}
                        experimentNames={this.state.uniqExperimentNames}
                        onExperimentChange={this.handleExperimentFilterChange}
                    />
                </div>
            </div>
            <div className="run-status-container">
                <div className="run-status-card-container">
                    <RunStatusCardContainer
                        runStatus={this.state.runStatistics}
                        onStatusFilterChange={this.handleStatusFilterChange} />
                </div>
                <div className="run-status-chart-container">
                    {<></> /*
                        <RunHistoryChart
                            metric={this.state.runHistoryChartData}
                            theme={this.context.theme}
                            configOverride={{
                                layout: {
                                    margin: {
                                        l: 20,
                                        r: 20,
                                        b: 30,
                                        t: 20
                                    },
                                },
                                config: {
                                    responsive: true,
                                    displayModeBar: false
                                }
                            }} />
                        */}
                </div>
            </div>
            <DataTable
                items={this.state.filteredRuns || []}
                itemsPerPage={15}
                columns={[
                    { field: "experimentName", header: "Experiment", minWidth: 70, maxWidth: 250 },
                    { field: "runId", header: "Run Id", minWidth: 70, maxWidth: 350, render: this.renderRunId },
                    { field: "status", header: "Status", minWidth: 80, maxWidth: 80 },
                    { field: "createdUtc", header: "Created", minWidth: 80, maxWidth: 200 },
                    { field: "startTimeUtc", header: "Start Time", minWidth: 80, maxWidth: 200 },
                    { field: "endTimeUtc", header: "End Time", minWidth: 80, maxWidth: 200 }
                ]}
                sortColumnFieldName="createdUtc"
                sortDescending={true}
            />
        </>
        );
    }

    protected readonly getData = async () => {
        const [runs, allowCreateExp] = await Promise.all([this.services.runHistoryService.getRunList(), this.services.accessService.checkPermission()]);

        if (!runs || allowCreateExp === undefined) {
            return;
        }
        const runHistoryChartData = processRunHistoryChart(runs);
        const uniqExperimentNames = this.getUniqExperimentNames(runs);
        const originRunStatistics = countBy(runs, (run) => run && run.status);
        const runStatistics = mergeRunStatus(originRunStatistics);
        if (allowCreateExp) {
            const startRunButton = this.navigationBarButtons.find((b) => b.key === "startRun");
            if (startRunButton) {
                startRunButton.disabled = false;
                this.refreshButtons();
            }
        }
        this.setState({ runs, uniqExperimentNames, runStatistics, runHistoryChartData }, this.handleDateExperimentChange);
    }
    private readonly startNewRun = () => {
        this.logUserAction("CreateExperiment");
        this.setState({
            redirectToStartRun: true
        });
    }

    private readonly renderRunId = (content: React.ReactNode, run: IRunDtoWithExperimentName) => {
        if (run.runId) {
            if (run.parentRunId) {
                return PageRedirectLink(content, ChildRun, { runId: run.runId, experimentName: run.experimentName });
            }
            else {
                return PageRedirectLink(content, ParentRun, { runId: run.runId, experimentName: run.experimentName });
            }
        }
        return undefined;
    }

    private readonly getUniqExperimentNames = (runs: IRunDtoWithExperimentName[]) => {
        const allExperimentNames = map(runs, (run) => run.experimentName);
        const uniqExperimentNames = uniq(allExperimentNames);
        uniqExperimentNames.sort();
        return uniqExperimentNames;
    }

    private readonly handleDateFilterChange = (startDate: Date | undefined) => {
        this.logUserAction("DateFilter", {
            DateFilterValue: startDate && moment(startDate)
                .utc()
                .toString()
        });
        this.setState({ dateFilter: startDate },
            this.handleDateExperimentChange);
    }

    private readonly handleExperimentFilterChange = (experimentFilter: string[] | undefined) => {
        this.logUserAction("ExperimentFilter", { ExperimentFilterValue: JSON.stringify(experimentFilter) });
        this.setState({ experimentFilter }, this.handleDateExperimentChange);
    }

    private readonly handleStatusFilterChange = (statusFilter: string | undefined) => {
        this.logUserAction("StatusFilter", { StatusFilterValue: statusFilter });
        this.setState({ statusFilter }, this.updateByStatusFilter);
    }

    private readonly updateRunListStatistic = () => {
        if (this.state.runs) {
            const filteredByDateExperimentRuns = this.state.runs.filter((run) =>
                this.filterRunByDate(run)
                && this.filterRunByExperiments(run));
            const originRunStatistics = countBy(filteredByDateExperimentRuns, (run) => run && run.status);
            const runStatistics = mergeRunStatus(originRunStatistics);
            this.setState({ runStatistics });
        }
    }

    private readonly updateByStatusFilter = () => {
        if (this.state.runs) {
            const filteredRuns = this.state.runs.filter((run) =>
                this.filterRunByDate(run)
                && this.filterRunByExperiments(run)
                && this.filterRunByStatus(run));
            const runHistoryChartData = processRunHistoryChart(filteredRuns);
            this.setState({ filteredRuns, runHistoryChartData });
        }
    }

    private readonly handleDateExperimentChange = async () => {
        this.updateRunListStatistic();
        this.updateByStatusFilter();
    }

    private readonly filterRunByDate = (run: IRunDtoWithExperimentName | undefined) => {
        if (!run || !this.state || !this.state.dateFilter) {
            return true;
        }
        return run.createdUtc && run.createdUtc >= this.state.dateFilter;
    }

    private readonly filterRunByExperiments = (run: IRunDtoWithExperimentName | undefined) => {
        if (!run || !this.state || !this.state.experimentFilter) {
            return true;
        }
        return includes(this.state.experimentFilter, run.experimentName);
    }

    private readonly filterRunByStatus = (run: IRunDtoWithExperimentName | undefined) => {
        if (!run || !this.state || !this.state.statusFilter) {
            return true;
        }
        return run.status && this.state.statusFilter === statusMapping[run.status];
    }
}
