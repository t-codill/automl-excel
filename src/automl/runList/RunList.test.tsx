import { shallow, ShallowWrapper } from "enzyme";
import { ICommandBarItemProps } from "office-ui-fabric-react";
import * as React from "react";
import { getLogCustomEventSpy, testContext } from "../common/context/__data__/testContext";
import { BaseComponent } from "../components/Base/BaseComponent";
import { DataTable } from "../components/DataTable/DataTable";
import { PageRedirectRender } from "../components/Redirect/PageRedirectRender";
import { AccessService } from "../services/AccessService";
import { IRunDtoWithExperimentName, RunHistoryService } from "../services/RunHistoryService";
import { IRunListState, RunList } from "./RunList";
import { RunListDateSelector } from "./RunListDateSelector";
import { RunListExperimentSelector } from "./RunListExperimentSelector";
import { RunStatusCardContainer } from "./RunStatusCardContainer";

jest.mock("../services/RunHistoryService");
jest.mock("../services/AccessService");
describe("RunList", () => {
    let getRunListSpy: jest.SpyInstance<
        ReturnType<RunHistoryService["getRunList"]>
    >;
    beforeEach(() => {
        getRunListSpy = jest.spyOn(RunHistoryService.prototype, "getRunList");
    });
    it("should show spinning", async () => {
        getRunListSpy.mockReturnValue(Promise.resolve(undefined));
        const tree = shallow<RunList>(<RunList />);
        await Promise.resolve();
        await Promise.resolve();
        expect(tree)
            .toMatchInlineSnapshot("<pageLoadingSpinner />");
    });
    it("should render", async () => {
        const tree = shallow<RunList>(<RunList />);
        await Promise.resolve();
        await Promise.resolve();
        expect(tree)
            .toMatchSnapshot();
    });
    it("should redirect to welcome page", async () => {
        getRunListSpy.mockReturnValue(Promise.resolve([]));
        const tree = shallow<RunList>(<RunList />);
        await Promise.resolve();
        await Promise.resolve();
        expect(tree
            .find(PageRedirectRender)
            .props())
            .toEqual({
                expendedRoutePath: "Welcome",
                noPush: false
            });
    });

    describe("start run button", () => {
        let checkPermissionSpy: jest.SpyInstance<
            ReturnType<AccessService["checkPermission"]>
        >;
        let startRunButton: ICommandBarItemProps | undefined;
        let tree: ShallowWrapper<{}, IRunListState>;
        beforeEach(async () => {
            checkPermissionSpy = jest.spyOn(
                AccessService.prototype,
                "checkPermission"
            );
            const setButtonSpy = jest.spyOn(BaseComponent.prototype.context, "setNavigationBarButtons");
            setButtonSpy.mockImplementation((buttons: ICommandBarItemProps[]) => {
                startRunButton = buttons.find((b) => b.key === "startRun");
            });
            tree = shallow<RunList>(<RunList />);
            await Promise.resolve();
            await Promise.resolve();
        });
        it("should show when have access", () => {
            let disabled: boolean | undefined;
            if (startRunButton) {
                disabled = startRunButton.disabled;
            }
            expect(disabled)
                .toBe(false);
        });
        it("should disable when no access", async () => {
            checkPermissionSpy.mockReturnValue(Promise.resolve(false));
            tree = shallow<RunList>(<RunList />);
            await Promise.resolve();
            await Promise.resolve();
            let disabled: boolean | undefined;
            if (startRunButton) {
                disabled = startRunButton.disabled;
            }
            expect(disabled)
                .toBe(true);
        });
        it("should not touch button if not found", async () => {
            checkPermissionSpy.mockImplementation(async () => {
                if (startRunButton) {
                    startRunButton.key = "startRunBak";
                    startRunButton.disabled = true;
                }
                return true;
            });
            tree = shallow<RunList>(<RunList />);
            await Promise.resolve();
            await Promise.resolve();
            let disabled: boolean | undefined;
            if (startRunButton) {
                disabled = startRunButton.disabled;
            }
            expect(disabled)
                .toBe(true);
        });
        it("should redirect to start run page", async () => {
            const logSpy = getLogCustomEventSpy();
            if (startRunButton && startRunButton.onClick) {
                startRunButton.onClick();
            }
            expect(logSpy)
                .toBeCalledWith(
                    "RunList_CreateExperiment_UserAction",
                    testContext,
                    { component: "CreateExperiment", pageName: "RunList" }
                );
            await Promise.resolve();
            expect(tree
                .find(PageRedirectRender)
                .props())
                .toEqual({
                    expendedRoutePath: "startRun",
                    noPush: false
                });
        });
    });

    describe("handleDateFilterChange", () => {
        const laterRun = {
            experimentName: "Experiment",
            createdUtc: new Date(12345)
        };
        const earlierRun = {
            experimentName: "Experiment",
            createdUtc: new Date(1234)
        };
        const sameTimeRun = {
            experimentName: "Experiment",
            createdUtc: new Date(10000)
        };
        const runs = [laterRun, earlierRun, sameTimeRun];
        let onDateChange: (d: Date) => void;
        let tree: ShallowWrapper<{}, IRunListState>;
        beforeAll(async () => {
            getRunListSpy.mockReturnValue(Promise.resolve(runs));
            tree = shallow<RunList>(<RunList />);
            await Promise.resolve();
            await Promise.resolve();
            onDateChange = tree.find(RunListDateSelector)
                .prop("onDateChange");
        });
        it("should filter", async () => {
            expect(tree.state("filteredRuns"))
                .toEqual(runs);
            const logSpy = getLogCustomEventSpy();
            onDateChange(new Date(10000));
            expect(logSpy)
                .toBeCalledWith(
                    "RunList_DateFilter_UserAction",
                    testContext,
                    { DateFilterValue: "Thu Jan 01 1970 00:00:10 GMT+0000", component: "DateFilter", pageName: "RunList" }
                );
            expect(tree.state("filteredRuns"))
                .toEqual([laterRun, sameTimeRun]);
        });
        it("should not filter if runs are undefined", async () => {
            tree.setState({ runs: undefined });
            const logSpy = getLogCustomEventSpy();
            onDateChange(new Date(20000));
            expect(logSpy)
                .toBeCalledWith(
                    "RunList_DateFilter_UserAction",
                    testContext,
                    { DateFilterValue: "Thu Jan 01 1970 00:00:20 GMT+0000", component: "DateFilter", pageName: "RunList" }
                );
            expect(tree.state("filteredRuns"))
                .toEqual([laterRun, sameTimeRun]);
        });
    });

    describe("handleExperimentFilterChange", () => {
        const runs = [
            {
                experimentName: "Experiment1",
                status: "Completed"
            },
            {
                experimentName: "Experiment2",
                status: "Running"
            },
            {
                experimentName: "Experiment3",
                status: "Cancelled"
            }
        ];
        let onExperimentChange: (d: string[]) => void;
        let tree: ShallowWrapper<{}, IRunListState>;
        beforeAll(async () => {
            getRunListSpy.mockReturnValue(Promise.resolve(runs));
            tree = shallow<RunList>(<RunList />);
            await Promise.resolve();
            await Promise.resolve();
            onExperimentChange = tree
                .find(RunListExperimentSelector)
                .prop("onExperimentChange");
        });

        it("should filter", async () => {
            expect(tree.state("filteredRuns"))
                .toEqual(runs);
            const logSpy = getLogCustomEventSpy();
            onExperimentChange(["Experiment1", "Experiment4"]);
            expect(logSpy)
                .toBeCalledWith(
                    "RunList_ExperimentFilter_UserAction",
                    testContext,
                    { ExperimentFilterValue: "[\"Experiment1\",\"Experiment4\"]", component: "ExperimentFilter", pageName: "RunList" }
                );
            expect(tree.state("filteredRuns"))
                .toEqual([
                    {
                        experimentName: "Experiment1",
                        status: "Completed"
                    }
                ]);
            expect(tree.state("runStatistics"))
                .toEqual({
                    Running: 0,
                    Completed: 1,
                    Failed: 0,
                    Others: 0
                });
        });
        it("should not filter if runs are undefined", async () => {
            tree.setState({ runs: undefined });
            const logSpy = getLogCustomEventSpy();
            onExperimentChange(["not exist"]);
            expect(logSpy)
                .toBeCalledWith(
                    "RunList_ExperimentFilter_UserAction",
                    testContext,
                    { ExperimentFilterValue: "[\"not exist\"]", component: "ExperimentFilter", pageName: "RunList" }
                );
            expect(tree.state("filteredRuns"))
                .toEqual([
                    {
                        experimentName: "Experiment1",
                        status: "Completed"
                    }
                ]);
        });
    });

    describe("handleStatusFilterChange", () => {
        const runs = [
            {
                experimentName: "Experiment1",
                status: "Completed"
            },
            {
                experimentName: "Experiment2",
                status: "Running"
            },
            {
                experimentName: "Experiment3",
                status: "Cancelled"
            }
        ];
        let onStatusFilterChange: (d: string) => void;
        let tree: ShallowWrapper<{}, IRunListState>;
        beforeAll(async () => {
            getRunListSpy.mockReturnValue(Promise.resolve(runs));
            tree = shallow<RunList>(<RunList />);
            await Promise.resolve();
            await Promise.resolve();
            onStatusFilterChange = tree
                .find(RunStatusCardContainer)
                .prop("onStatusFilterChange");
        });

        it("should filter", async () => {
            expect(tree.state("filteredRuns"))
                .toEqual(runs);
            const logSpy = getLogCustomEventSpy();
            onStatusFilterChange("Completed");
            expect(logSpy)
                .toBeCalledWith(
                    "RunList_StatusFilter_UserAction",
                    testContext,
                    { StatusFilterValue: "Completed", component: "StatusFilter", pageName: "RunList" }
                );
            expect(tree.state("filteredRuns"))
                .toEqual([
                    {
                        experimentName: "Experiment1",
                        status: "Completed"
                    }
                ]);
        });
        it("should not filter if runs are undefined", async () => {
            tree.setState({ runs: undefined });
            const logSpy = getLogCustomEventSpy();
            onStatusFilterChange("not exist");
            expect(logSpy)
                .toBeCalledWith(
                    "RunList_StatusFilter_UserAction",
                    testContext,
                    { StatusFilterValue: "not exist", component: "StatusFilter", pageName: "RunList" }
                );
            expect(tree.state("filteredRuns"))
                .toEqual([
                    {
                        experimentName: "Experiment1",
                        status: "Completed"
                    }
                ]);
        });
    });

    describe("renderRunId", () => {
        let render: (
            content: React.ReactNode,
            data: IRunDtoWithExperimentName
        ) => React.ReactNode;
        beforeEach(async () => {
            const tree = shallow<RunList>(<RunList />);
            await Promise.resolve();
            await Promise.resolve();
            render = tree
                .find(DataTable)
                .prop("columns")[1].render as (
                    content: React.ReactNode,
                    data: IRunDtoWithExperimentName
                ) => React.ReactNode;
        });
        it("should return undefined without run id", async () => {
            expect(
                render(<div>Content</div>, {
                    experimentName: "ExpName"
                })
            )
                .toBeUndefined();
        });
        it("should return parent run link", async () => {
            expect(
                render(<div>Content</div>, {
                    experimentName: "ExpName",
                    runId: "runId"
                })
            )
                .toMatchInlineSnapshot(`
<PageRedirectLinkRender
  expendedRoutePath="experiments/ExpName/parentrun/runId"
>
  <div>
    Content
  </div>
</PageRedirectLinkRender>
`);
        });
        it("should return child run link", async () => {
            expect(
                render(<div>Content</div>, {
                    experimentName: "ExpName",
                    runId: "runId",
                    parentRunId: "parentRunId"
                })
            )
                .toMatchInlineSnapshot(`
<PageRedirectLinkRender
  expendedRoutePath="experiments/ExpName/childrun/runId"
>
  <div>
    Content
  </div>
</PageRedirectLinkRender>
`);
        });
    });
});
