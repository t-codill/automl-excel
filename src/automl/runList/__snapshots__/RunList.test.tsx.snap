// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RunList should render 1`] = `
<Fragment>
  <div
    className="run-filter-container"
  >
    <div
      className="run-filter-action-container"
    >
      <RunListDateSelector
        onDateChange={[Function]}
      />
    </div>
    <div
      className="run-filter-action-container"
    >
      <RunListExperimentSelector
        experimentNames={
          Array [
            "experiment1",
            "experiment2",
          ]
        }
        isAllSelect={true}
        onExperimentChange={[Function]}
      />
    </div>
  </div>
  <div
    className="run-status-container"
  >
    <div
      className="run-status-card-container"
    >
      <RunStatusCardContainer
        onStatusFilterChange={[Function]}
        runStatus={
          Object {
            "Completed": 0,
            "Failed": 0,
            "Others": 0,
            "Running": 0,
            "undefined": NaN,
          }
        }
      />
    </div>
    <div
      className="run-status-chart-container"
    >
      <RunHistoryChart
        configOverride={
          Object {
            "config": Object {
              "displayModeBar": false,
              "responsive": true,
            },
            "layout": Object {
              "margin": Object {
                "b": 30,
                "l": 20,
                "r": 20,
                "t": 20,
              },
            },
          }
        }
        metric={
          Object {
            "1970/01/01": Object {
              "undefined": 2,
            },
          }
        }
        theme="light"
      />
    </div>
  </div>
  <DataTable
    columns={
      Array [
        Object {
          "field": "experimentName",
          "header": "Experiment",
          "maxWidth": 250,
          "minWidth": 70,
        },
        Object {
          "field": "runId",
          "header": "Run Id",
          "maxWidth": 350,
          "minWidth": 70,
          "render": [Function],
        },
        Object {
          "field": "status",
          "header": "Status",
          "maxWidth": 80,
          "minWidth": 80,
        },
        Object {
          "field": "createdUtc",
          "header": "Created",
          "maxWidth": 200,
          "minWidth": 80,
        },
        Object {
          "field": "startTimeUtc",
          "header": "Start Time",
          "maxWidth": 200,
          "minWidth": 80,
        },
        Object {
          "field": "endTimeUtc",
          "header": "End Time",
          "maxWidth": 200,
          "minWidth": 80,
        },
      ]
    }
    items={
      Array [
        Object {
          "experimentName": "experiment1",
          "runId": "run1",
        },
        Object {
          "experimentName": "experiment2",
          "runId": "run2",
        },
      ]
    }
    itemsPerPage={15}
    sortColumnFieldName="createdUtc"
    sortDescending={true}
  />
</Fragment>
`;
